### PR TITLE
add check for system assemblies completion

### PR DESCRIPTION
**Pull request by dsyme**
The list of things we consider "system assemblies" was not complete for .NET 5.0, and most importantly didn't contain System.Net.WebClient, which causes failures in using the arguments generated by GetProjectOptionsFromScript when run as part of a .NET 5.0 application.

### CODE REVIEW

File: src/fsharp/DotNetFrameworkDependencies.fs [hunk 1]
```diff
@@ -482,6 +482,162 @@ module internal FSharp.Compiler.DotNetFrameworkDependencies
...
+            yield "System.Xml"
+            yield "System.Xml.Linq"
+            yield "System.Xml.ReaderWriter"
+            yield "System.Xml.Serialization"
+            yield "System.Xml.XDocument"
+            yield "System.Xml.XmlDocument"
+            yield "System.Xml.XmlSerializer"
+            yield "System.Xml.XPath"
+            yield "System.Xml.XPath.XDocument"
+            yield "WindowsBase"
```
Comment (baronfel) [inline src/fsharp/DotNetFrameworkDependencies.fs:640]: Are WindowsBase/System.Windows/Microsoft.Win32.Primitives valid for non-windows FSI sessions?

Similarly, should all three 'primary assemblies' of mscorlib, System.Runtime, and netstandard.dll be referenced?

**Comment by dsyme** [comment]
> Are WindowsBase/System.Windows/Microsoft.Win32.Primitives valid for non-windows FSI sessions?

`systemAssemblies` is not a list of references.. Instead this is a list of things which, *if referenced*, go in the shared "tcFrameworkImports" which is shared between other TcImports.

Putting something in this list has two effects

1. It is a candidate for the shared tcFrameworkImports, hence more sharing may happen

2. It is reference is processed before everything else.

The specific problem solved here is that `System.Net.WebClient` should be in this list, because FSharp.Core tries to resolve a reference into this assembly.  Hence its reference should be processed before everything else

File: src/fsharp/DotNetFrameworkDependencies.fs (hunk 1 above)
Comment (dsyme) [reply src/fsharp/DotNetFrameworkDependencies.fs]: see response below, thanks, it's a good question

**Comment by KevinRansom** [comment]
@dsyme 
I'm assuming that you got the list of assemblies to add to the sdkAssemblies from the list in the packs directory which, is fine.  Having the sdkAssemblies table contain each framework list is somewhat duplicative.  It for sure means a longer scan through the source for us to to figure out if a particular assembly is in the list.  I imagine that formatting it in alphabetical sorted order with each assembly only once would be more practical.

@dsyme the thing that I am most curious about is the FSharp.Core dependency on System.Web.Client.  As far as I can tell neither fsi, fsharp.compiler.private nor fsharp.core depend on it.  The reason it is in the set is that for use sdk refs, we specify the entire set of netsdk assemblies as references.

I would prefer, that rather than having this list, grow endlessly as the sdk grows, that we just add code to compute it similarly to what we do for usesdk and shrink the list back down again.

The test is good.

I will merge this, and consider a permanent solution.

**Review by KevinRansom — APPROVED** [review]

**Comment by dsyme** [comment]
> I'm assuming that you got the list of assemblies to add to the sdkAssemblies from the list in the packs directory which, is fine. Having the sdkAssemblies table contain each framework list is somewhat duplicative. It for sure means a longer scan through the source for us to to figure out if a particular assembly is in the list. I imagine that formatting it in alphabetical sorted order with each assembly only once would be more practical.

TBH I got the lists from the respective results of GetProjectOptionsFromScript.

> @dsyme the thing that I am most curious about is the FSharp.Core dependency on System.Web.Client. As far as I can tell neither fsi, fsharp.compiler.private nor fsharp.core depend on it. The reason it is in the set is that for use sdk refs, we specify the entire set of netsdk assemblies as references.

I think the references go via netstandard.dll

In .NET 5.0,  netstandard.dll (`"c:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\5.0.0\ref\net5.0\netstandard.dll") has this:

```
.assembly extern System.Net.WebClient
{
  .publickeytoken = (CC 7B 13 FF CD 2D DD 51 )                         // .{...-.Q
  .ver 5:0:0:0
}
```


> I would prefer, that rather than having this list, grow endlessly as the sdk grows, that we just add code to compute it similarly to what we do for usesdk and shrink the list back down again.

Yup

> I will merge this, and consider a permanent solution.

cool

**Comment by dsyme** [comment]
@baronfel If you could publish a new FCS I'd be really grateful, when you have a chance

**Comment by KevinRansom** [comment]
@dsyme publishing fcs is on us now.

**Comment by KevinRansom** [comment]
The code needs to flow to dev16 branch, then I will push a new fcs.

**Comment by auduchinok** [comment]
@KevinRansom What's the connection between dev16 branch and publishing new FCS version from `main`?

**Comment by KevinRansom** [comment]
@auduchinok, we originally were going to publish everything from the active release branch.  However ... that would require the dogfood build, which I realized once I had thought about it a bit.  We haven't really got the hang of releasing fcs yet, so we haven't identified the correct process.  TBH my preference would be releasing from a servicing branch with only cherry picked features coming through.  Because as pointed out by @baronfel elsewhere API deletions have already occured which requires a major version number change.  Given that those type of changes happen in master continuously anyway .. more thought is required for the actual get it out the door step.

In the meantime, we will ship this one from master and try to figure out what to do in the future.

**Comment by auduchinok** [comment]
@KevinRansom I think it'll be really great if changes regarding publishing the community bits are announced clearly. It's been published from `master`/`main` previously, and if it's going to be changed to, say, using some stable/outdated branch, it really should be discussed somewhere, since many community projects tend to rely on recent changes in FCS.

**Comment by KevinRansom** [comment]
@auduchinok well, ideally we would preview from release/dev16.9 and service from release/16.8.  So that we could protect most developers from churn and not have a major version number that rotated faster than a rocket.  On the other hand, it is work.

**Comment by baronfel** [comment]
The community (speaking for FSAC, fantomas, fsharplint, and FAKE here since those are the ones I generally maintain) is generally ok with rapidly-increasing version numbers, so long as we get something for it ;)  It's just a number, and I think we much more value rapid delivery over something cosmetic like that.

**Comment by KevinRansom** [comment]
@baronfel --- I think a reasonable approach to take is to publish previews out of main branch.  I.e. 39.0.0-betasomecrazynumber
And have a servicing branch for fcsserving to publish servicing fixes to 38.0.n with fixes cherry picked into fcs servicing that don't break apis.

@brettfo , @cartermp what do you think, will this work?

**Comment by KevinRansom** [comment]
We can always modify this proposal if it doesn't work out to be helpful.

**Pull request merged**